React renderer for MCP Apps

ochafik opened the pull request:
AppRenderer is a renderer for tool calls that return MCP Apps, adapted [from this example](https://github.com/modelcontextprotocol/ext-apps/blob/main/examples/simple-host/src/AppRenderer.tsx) to accept an optional MCP-UI `onUIAction` callback.

Apps can be created with the SDK at https://github.com/modelcontextprotocol/ext-apps

More details on [SEP-1865](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1865)

Requires updating MCP TS SDK from 1.11 to 1.22: sent separately as https://github.com/MCP-UI-Org/mcp-ui/pull/148, merged here

cc/ @idosal @liady @antonpk1

infoxicator commented on pnpm-lock.yaml:
```diff
@@ -251,8 +251,8 @@ importers:
-        specifier: ^1.11.1
-        version: 1.13.1
+        specifier: ^1.22.0
+        version: 1.22.0
```
```suggestion
        version: 1.16.0
```

infoxicator commented on pnpm-lock.yaml:
```diff
@@ -251,8 +251,8 @@ importers:
+        version: 1.22.0
       '@vis.gl/react-google-maps':
         specifier: ^1.5.2
         version: 1.5.3(react-dom@19.1.0(react@19.1.0))(react@19.1.0)
```
```suggestion
        version: 1.5.3(react-dom@19.2.0(react@19.2.0))(react@19.2.0)
```

infoxicator commented on pnpm-lock.yaml:
```diff
@@ -1913,13 +1925,18 @@ packages:
+      '@cfworker/json-schema': ^4.1.1
+    peerDependenciesMeta:
+      '@cfworker/json-schema':
+        optional: true
```
```suggestion

```

infoxicator commented on pnpm-lock.yaml:
```diff
@@ -11702,13 +11846,24 @@ snapshots:
       fast-json-stable-stringify: 2.1.0
       json-schema-traverse: 0.4.1
       uri-js: 4.4.1
 
```
```suggestion
      ai: 4.3.16(react@19.2.0)(zod@3.25.67)
```

infoxicator commented on pnpm-lock.yaml:
```diff
@@ -11702,13 +11846,24 @@ snapshots:
+    dependencies:
+      fast-deep-equal: 3.1.3
+      fast-uri: 3.1.0
+      json-schema-traverse: 1.0.0
```
```suggestion
      react: 19.2.0
```

idosal commented on sdks/typescript/client/src/components/AppRenderer.tsx:
```diff
@@ -0,0 +1,389 @@
+ */
+export interface AppRendererProps {
+  /** URL to the sandbox proxy HTML that will host the tool UI iframe */
+  sandboxProxyUrl: URL;
```
Perhaps we should change it to a `sandbox` prop with an internal object for `url`. I imagine `sandbox` will have additional properties in the future (such as permissions). WDYT?

idosal commented on sdks/typescript/client/src/components/AppRenderer.tsx:
```diff
@@ -0,0 +1,389 @@
+  sandboxProxyUrl: URL;
+
+  /** MCP client connected to the server providing the tool */
+  client: Client;
```
@infoxicator @ochafik Can we extract this component from `AppRenderer` so we'll export both the bare-bones component and the higher-level renderer on top of it?
I can take a stab at it later.

chelojimenez replied:
hi @ochafik, we built our own [renderer](https://github.com/MCPJam/inspector/blob/main/client/src/components/chat-v2/mcp-apps-renderer.tsx) in MCPJam to provide initial support of MCP Apps, this renderer like @infoxicator points out, we pass HTML directly rather than requiring an MCP client instance (This pattern works well when the MCP client lives in a different context and ours lives in the server). Happy to discuss any of these if useful!

liady replied:
@chelojimenez yeah, it's much simpler if the host can extract the HTML itself and pass it, and I agree that in your case it's even the only way (since the rendering context doesn't have an access to the client instance).
Note that @ochafik 's usage here is much wider - since this implementation relies on the `AppBridge` (from the [ext-apps repo](https://github.com/modelcontextprotocol/ext-apps/blob/main/src/app-bridge.ts)) - which needs the client in order to proxy MCP requests.
In the MCPJam implementation this is not needed - [mcp-apps-renderer](https://github.com/MCPJam/inspector/blob/main/client/src/components/chat-v2/mcp-apps-renderer.tsx) actually doesn't rely on MCP at all - and simply parses messages according to the schema. (It's a good renderer btw, even if it's not "generic" in that sense).

It's a good question regarding how we want to build the **generic** client side renderer here. @ochafik 's implementation encapsulates a lot of the functionality (including HTML resource mathcing) - but requires the `client` instance. On the other hand a more "low-level" renderer would leave more heavy-lifting to the host, and if we skip using AppBridge we might be missing the MCP re-use.

As @idosal is also working on the client side SDK, we'll probably need to align on the best abstraction here. I think that allowing (optionally) to pass the raw resource HTML instead of the entire client is a good, but we still need to solve for the usage of `AppBridge` that does require the client instance.

ochafik replied:
Addressed in d12ddcc with two approaches:

**Option 1: Use `AppRenderer` with `html` prop** (still get AppBridge/MCP support):
```tsx
<AppRenderer
  client={mcpClient}
  toolName="my-tool"
  sandbox={{ url: sandboxUrl }}
  html={preloadedHtml}  // skips all resource fetching
  toolInput={args}
/>
```

**Option 2: Use `AppFrame` directly** (no client required):
```tsx
<AppFrame
  html={myHtml}
  sandbox={{ url: sandboxUrl }}
  onOpenLink={(url) => window.open(url)}
  onMessage={(content) => console.log(content)}
  onSizeChange={({ width, height }) => ...}
  onLoggingMessage={(params) => ...}
  onInitialized={({ appVersion, appCapabilities }) => ...}
  toolInput={args}
/>
```

`AppFrame` works standalone for hosts like MCPJam where the MCP client lives in a different context. Pass an `appBridge` prop if you need full MCP proxying.

cc @infoxicator @chelojimenez

guru3s commented on sdks/typescript/client/src/utils/app-host-utils.ts:
```diff
@@ -0,0 +1,114 @@
+        if (
+          event.data &&
+          event.data.method ===
+            McpUiSandboxProxyReadyNotificationSchema.shape.method._def.value
```
this should be .values[0]

ochafik replied:
✅ Fixed in e0c4421 — good catch on the Zod v4 API change!

aharvard commented:
This is great! Looking forward to using it in Goose when ready.

ochafik commented:
## Update: ext-apps v0.2.0 upgrade

Upgraded `@modelcontextprotocol/ext-apps` to `^0.2.0` and `@modelcontextprotocol/sdk` to `^1.24.0`.

### Breaking Changes in ext-apps v0.2.0

1. **Method renaming** (with deprecated aliases):
   - `sendResourceTeardown()` → `teardownResource()`
   - `sendOpenLink()` → `openLink()` (not used in mcp-ui)

2. **MCP SDK moved to peer dependency** - requires `@modelcontextprotocol/sdk@^1.24.0`

3. **Optional transport in `connect()`** - can now call without parameters (auto-detects platform)

### Changes in mcp-ui

- Updated all packages to use `^0.2.0` instead of `github:...#main`
- Renamed `AppRendererHandle.sendResourceTeardown` → `teardownResource` to align with upstream
- Updated MCP SDK to `^1.24.0` across all packages

### Additional fixes

- Added documentation for `AppRenderer` and `AppFrame` components in mcp-apps.md
- Fixed promise rejection handling in `setupSandboxProxyIframe` (timeout + error listener)

All tests pass ✅

infoxicator commented on sdks/typescript/client/package.json:
```diff
@@ -19,11 +19,14 @@
-    "@modelcontextprotocol/sdk": "^1.22.0",
+    "@modelcontextprotocol/ext-apps": "^0.2.0",
+    "@modelcontextprotocol/sdk": "^1.24.0",
+    "@mcp-ui/shared": "workspace:*",
```
@ochafik was this change needed? it is causing a resolution issue when trying to install locally `➤ YN0001: │ Error: @mcp-ui/shared@workspace:*: Workspace not found (@mcp-ui/shared@workspace:*)` removing this line fixes the error

infoxicator commented on sdks/typescript/client/src/components/AppFrame.tsx:
```diff
@@ -0,0 +1,253 @@
+      mounted = false;
+      if (iframeRef.current && containerRef.current?.contains(iframeRef.current)) {
+        containerRef.current.removeChild(iframeRef.current);
+      }
```
```suggestion

```

@ochafik found a lifecycle problem. This code destroys the iframe when the component unmounts. but then the mounting process doesn't happen again (create iframe setHtml). I don't think that's what we want anyway?

The iframe should preserve the state when rerendering instead of destroying and recreating the iframe?

idosal commented:
@ochafik - the eslint update on `main` caused conflicts with the PR and revealed a few issues. Fixed them all and merged

infoxicator commented on sdks/typescript/client/src/components/AppFrame.tsx:
```diff
@@ -0,0 +1,287 @@
+        // Register logging handler
+        appBridge.onloggingmessage = (params) => {
+          onLoggingMessageRef.current?.(params);
+        };
```
```suggestion

```
this is setting the logging callback handler twice and is overriding the one passed in the AppRenderer

making the callback not to fire.

ochafik commented:
> @ochafik - updated the proxy ready method name to use the new exported const from ext-apps. It uses the PR package to unblock alpha testers, but I'll switch it back to the official version after the ext-apps PR is merged.

@idosal Not sure what happened to that const, i've defined it locally (in a typesafe way referring to the message) along w/ a few tweaks / using new utils.

idosal commented:
> Gentle ping @idosal @liady , anything else needed here?

Was waiting for you 😄

Thanks a lot for contributing this critical capability @ochafik, you rock! Special thanks to @infoxicator and the community for helping push this through. ❤️

infoxicator commented:
> > Gentle ping @idosal @liady , anything else needed here?
> 
> Was waiting for you 😄
> 
> Thanks a lot for contributing this critical capability @ochafik, you rock! Special thanks to @infoxicator and the community for helping push this through. ❤️

You all rock! thanks for pushing this 👏
